code/game/objects/items/RSF.dm

```diff
@@ -8,178 +8,203 @@ RSF
+	var/list/matter_by_item = list(/obj/item/rcd_ammo = 10,)
+	///A list of surfaces that we are allowed to place things on.
+	var/list/allowed_surfaces = list(/turf/open/floor, /obj/structure/table)
+	///The unit of mesure of the matter, for use in text
+	var/discriptor = "fabrication-units"
+	///The verb that describes what we're doing, for use in text
+	var/action_type = "Dispensing"
+
+/obj/item/rsf/Initialize()
+	. = ..()
+	to_dispense = cost_by_item[1]
 
 /obj/item/rsf/examine(mob/user)
 	. = ..()
-	. += "<span class='notice'>It currently holds [matter]/30 fabrication-units.</span>"
+	. += "<span class='notice'>It currently holds [matter]/[max_matter] [discriptor].</span>"
 
 /obj/item/rsf/cyborg
 	matter = 30
 
 /obj/item/rsf/attackby(obj/item/W, mob/user, params)
-	if(istype(W, /obj/item/rcd_ammo))
-		if((matter + 10) > 30)
-			to_chat(user, "<span class='warning'>The RSF can't hold any more matter!</span>")
+	if(is_type_in_list(W,matter_by_item))//If the thing we got hit by is in our matter list
+		var/tempMatter = matter_by_item[W.type] + matter
+		if(tempMatter > max_matter)
+			to_chat(user, "<span class='warning'>\The [src] can't hold any more [discriptor]!</span>")
 			return
 		qdel(W)
-		matter += 10
+		matter = tempMatter //We add its value
 		playsound(src.loc, 'sound/machines/click.ogg', 10, TRUE)
-		to_chat(user, "<span class='notice'>The RSF now holds [matter]/30 fabrication-units.</span>")
-		icon_state = "rsf"
+		to_chat(user, "<span class='notice'>\The [src] now holds [matter]/[max_matter] [discriptor].</span>")
+		icon_state = initial(icon_state)//and set the icon state to the initial value it had
 	else
 		return ..()
 
 /obj/item/rsf/attack_self(mob/user)
 	playsound(src.loc, 'sound/effects/pop.ogg', 50, FALSE)
-	switch(mode)
-		if(5)
-			mode = 1
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Drinking Glass'.</span>")
-		if(1)
-			mode = 2
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Paper'.</span>")
-		if(2)
-			mode = 3
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Pen'.</span>")
-		if(3)
-			mode = 4
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Dice Pack'.</span>")
-		if(4)
-			mode = 5
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Cigarette'.</span>")
+	var/atom/target = cost_by_item
+	var/cost = 0
+	//Warning, prepare for bodgecode
+	while(istype(target, /list/))//While target is a list we continue the loop
+		var/picked = show_radial_menu(user, src, formRadial(target), custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE)
+		if(!check_menu(user) || picked == null)
+			return
+		for(var/emem in target)//Back through target agian
+			var/atom/test = extractObject(target, emem)
+			if(picked == initial(test.name))//We try and find the entry that matches the radial selection
+				cost = target[emem]//We cash the cost
+				target = emem
+				break
+			qdel(test)
```

Comment:
No need to delete the atom now because it's not being initialized, correct?

Reply:
yeah......